ouroboros-consensus-cardano-tools/src/Cardano/Tools/ImmDBServer/MiniProtocols.hs

```diff
@@ -0,0 +1,229 @@
+     )
+  => CodecConfig blk
+  -> ImmutableDB m blk
+  -> NetworkMagic
+  -> Versions NodeToNodeVersion NodeToNodeVersionData
+       (OuroborosApplication 'ResponderMode addr BL.ByteString m Void ())
+immDBServer codecCfg immDB networkMagic = do
+    forAllVersions application
+  where
+    forAllVersions ::
+         (NodeToNodeVersion -> BlockNodeToNodeVersion blk -> r)
+      -> Versions NodeToNodeVersion NodeToNodeVersionData r
+    forAllVersions mkR =
+          Versions
+        $ Map.mapWithKey mkVersion
+        $ supportedNodeToNodeVersions (Proxy @blk)
+      where
+        mkVersion version blockVersion = Version {..}
+          where
+            versionApplication _ = mkR version blockVersion
+            versionData = NodeToNodeVersionData {..}
+              where
+                diffusionMode = N2N.InitiatorOnlyDiffusionMode
+
+    application ::
+         NodeToNodeVersion
+      -> BlockNodeToNodeVersion blk
+      -> OuroborosApplication 'ResponderMode addr BL.ByteString m Void ()
+    application version blockVersion =
+        OuroborosApplication \_connId _controlMessageSTM -> miniprotocols
+      where
+        miniprotocols =
+            [ mkMiniProtocol
+                N2N.keepAliveMiniProtocolNum
+                N2N.keepAliveProtocolLimits
+                keepAliveProt
+            , mkMiniProtocol
+                N2N.chainSyncMiniProtocolNum
+                N2N.chainSyncProtocolLimits
+                chainSyncProt
+            , mkMiniProtocol
+                N2N.blockFetchMiniProtocolNum
+                N2N.blockFetchProtocolLimits
+                blockFetchProt
+            , mkMiniProtocol
+                N2N.txSubmissionMiniProtocolNum
+                N2N.txSubmissionProtocolLimits
+                txSubmissionProt
+            ]
+          where
+            Consensus.N2N.Codecs {..} =
+              Consensus.N2N.defaultCodecs codecCfg blockVersion version
+
+            keepAliveProt  =
+                MuxPeer nullTracer cKeepAliveCodec
+              $ keepAliveServerPeer keepAliveServer
+            chainSyncProt  =
+                MuxPeerRaw \channel ->
+                withRegistry
+              $ runPeer nullTracer cChainSyncCodecSerialised channel
+              . chainSyncServerPeer
+              . chainSyncServer immDB getSerialisedHeader
+            blockFetchProt =
+                MuxPeerRaw \channel ->
+                withRegistry
+              $ runPeer nullTracer cBlockFetchCodecSerialised channel
+              . blockFetchServerPeer
+              . blockFetchServer immDB (Serialised <$> GetRawBlock)
+            txSubmissionProt =
+                MuxPeerRaw \_channel -> forever $ threadDelay 10
+
+
+        mkMiniProtocol miniProtocolNum limits proto = MiniProtocol {..}
+          where
+            miniProtocolLimits = limits N2N.defaultMiniProtocolParameters
+            miniProtocolRun    = ResponderProtocolOnly proto
+
+chainSyncServer ::
```

Comment:
> Ah, bummer we can't re-use the actual mini protocol definitions.

I think the differences between ChainDB `Follower`s and ImmutableDB `Iterator`s also *could* be abstracted over, but it would probably more messy/adhoc than for BlockFetch.

> It might be worth a comment explaining that a normal ChainSync client would only ever send FindIntersect more than once (for the initialization) if it was about to disconnect from this ImmDB server, but the CSJ we're currently prototyping will send them more than once.

Ah, the motivation for this comment would be that we can't just keep one iterator open with the CSJ use case in mind? Yeah, will add a comment :+1: